tools/lua_sstable_consumer.cc

```diff
@@ -0,0 +1,1003 @@
+        return 1;
+    } else if (!strcmp(field, "tombstone")) {
+        if (!rtc.tombstone()) {
+            return 0;
+        }
+        push_userdata<tombstone>(l, rtc.tombstone(), tombstone_metatable);
+        return 1;
+    }
+    return 0;
+}
+
+class json_writer {
+    using stream = rapidjson::BasicOStreamWrapper<std::ostream>;
+    using writer = rapidjson::Writer<stream, rjson::encoding, rjson::encoding, rjson::allocator>;
+
+    stream _stream;
+    writer _writer;
+
+private:
+    static json_writer& get_this(lua_State* l) {
+        return pop_userdata_as_ref<json_writer>(l, 1);
+    }
+
+    template <typename Function>
+    static int invoke(lua_State* l, const char* name, int expected_type, Function&& fun) {
+        auto nargs = int(expected_type != LUA_TNIL) + 1;
+        if (const auto n = lua_gettop(l); n != nargs) {
+            return luaL_error(l, "json_writer::%s(): expected %I arguments, got %I", name, nargs, n);
+        }
+        const auto type = lua_type(l, 2);
+        if (expected_type != LUA_TNIL && type != expected_type) {
+            return luaL_error(l, "json_writer::%s(): expected argument of type %s, got %s", name, lua_typename(l, expected_type), lua_typename(l, type));
+        }
+        try {
+            fun(get_this(l));
+        } catch (const std::exception& e) {
+            return luaL_error(l, "json_writer::%s(): %s", name, e.what());
+        }
+        return 0;
+    }
+
+public:
+    json_writer() : _stream(std::cout), _writer(_stream)
+    { }
+    static int null_l(lua_State* l) {
+        return invoke(l, __FUNCTION__, LUA_TNIL, [l] (json_writer& w) {
+            w._writer.Null();
+        });
+    }
+    static int bool_l(lua_State* l) {
+        return invoke(l, __FUNCTION__, LUA_TBOOLEAN, [l] (json_writer& w) {
+            w._writer.Bool(lua_toboolean(l, 2));
+        });
+    }
+    static int int_l(lua_State* l) {
+        return invoke(l, __FUNCTION__, LUA_TNUMBER, [l] (json_writer& w) {
+            w._writer.Int64(lua_tointeger(l, 2));
+        });
+    }
+    static int double_l(lua_State* l) {
+        return invoke(l, __FUNCTION__, LUA_TNUMBER, [l] (json_writer& w) {
+            w._writer.Double(lua_tonumber(l, 2));
+        });
+    }
+    static int string_l(lua_State* l) {
+        return invoke(l, __FUNCTION__, LUA_TSTRING, [l] (json_writer& w) {
+            size_t len = 0;
+            auto str = lua_tolstring(l, 2, &len);
+            w._writer.String(str, len, false);
+        });
+    }
+    static int start_object_l(lua_State* l) {
+        return invoke(l, __FUNCTION__, LUA_TNIL, [l] (json_writer& w) {
+            w._writer.StartObject();
+        });
+    }
+    static int key_l(lua_State* l) {
+        return invoke(l, __FUNCTION__, LUA_TSTRING, [l] (json_writer& w) {
+            size_t len = 0;
+            auto str = lua_tolstring(l, 2, &len);
+            w._writer.Key(str, len, false);
+        });
```

Comment:
What if the script changes key but not key_raw? We could end up with an inconsistency.

Probably key and key_taw should both be references to the same object, so when one is updated, so is the other. But this sort of subtle bug can be left there, since likely most uses will be to filter or accumulate data, not to change it.

Reply:
Currently all attributes are read-only. But in any case Lua implements attribute access via a separate get function and a separate set function. So we are required to intercept all write attempts to attributes and can easily implement internal requirement enforcing from there.